### ISSUE TITLE
AWSMobileClient race condition in calling confirmSignIn from sign callback

### ISSUE THREAD

**Issue by royjit** [aws-amplify/aws-sdk-ios#1686]
Originally reported here #1248 
**Describe the bug**
AWSMobileClient.confirmSignIn callback is not called when we invoke `AWSMobileClient.confirmSignIn` from `AWSMobileClient.signIn`'s callback.

**To Reproduce**

```
func signupScenario() {
       AWSMobileClient.sharedInstance().signIn(username: "testuser", password: "Test123!", validationData: nil) { signInResult, error in

           if let result = signInResult {
              
               switch result.signInState {
               case .newPasswordRequired:
                   self.setNewPassword("test123123")
               default:
                   break
               }
           } else if let customError = error {

           }
       }
   }

   func setNewPassword(_ password: String) {
       AWSMobileClient.sharedInstance().confirmSignIn(challengeResponse: password) { signInResult, error in
           // <-- this callback never gets called
           
       }
   }
```
**Which AWS service(s) are affected?** 
Cognito

**Expected behavior**
`AWSMobileClient.confirmSignIn`'s callback should be called

**Comment by palpatim** [comment]
Once we release a fix that resolves this issue, we need to ask @Skifcha to verify that it fixes #1669.

**Comment by palpatim** [comment]
We released this in https://github.com/aws-amplify/aws-sdk-ios/releases/2.11.0

Please let us know if you have any problems with this feature.
